cms_coding_rules.md

```diff
@@ -0,0 +1,144 @@
+# CMS Naming, Coding, And Style Rules
+
+### By the CMS Ofﬂine Software Development Team
+
+## Abstract
+This document describes the design, naming, coding, and style rules and recommendations for CMS software written in C++, plus a few guidelines for Python scripts.
+
+
+## Outline
+#### 1 -- Introduction
+#### 2 -- Naming Rules
+#### 3 --Style Rules
+#### 4 -- Technical Coding Rules
+#### 5 -- Documentation Rules
+#### 6 -- Packaging Rules
+#### 7 -- Design and Coding Guidelines
+
+## 1 -- Introduction
+This document describes the CMS C++ software naming, coding, style and documentation rules and recommendations.
+All CMS C++ software is expected to comply with the rules. The asterisk (\*) after some rules indicates that there may be exceptional use cases where the rule may be violated with good justiﬁcation.
+Coding rules are meant to prevent serious problems in software function, performance, maintainability, usability, and portability. The Packaging Rules section also has some brief guidelines for Python scripts.
+
+## 2 -- Naming Rules
+1. C++ header ﬁles use the sufﬁx .h, e.g. `CaloCluster.h` (*)
+2. For C++ source ﬁles, the preferred sufﬁx  is .cc, e.g. `CaloCluster.cc`, though there are existing files using .cpp and .cxx suffixes. (*)
+3. For a header file that contains a class, name that ﬁle after the class.
+4. Name source ﬁles after the class.
+5. For class, struct, type and enumeration names use upper class initials, e.g. `GeometryBuilder`.
+6. For namespaces use lower case, e.g. `namespace edm`.
+7. Start method names with lowercase, use upper case initials for following words, e.g. `collisionPoint()` (allowed exception: implementation of virtual methods inherited from external packages e.g. `ProcessHits()` method required by Geant4).
+8. Start data member names with lower case. A trailing “_” is the preferred method to distinguish a data member from the getter method (e.g. `momentum_`).
+9. Using “set” for a setter method is preferred, e.g. `setMomentum(double m)`.
+10. For a getter method, using the value name is preferred, e.g. `momentum()`.
+11. Do not use single character names, except for loop indices.
+12. Do not use special characters, except for “_” where allowed.
+13. Do not use “_” as ﬁrst character. Only use it as the last character for class data member names, not local variable names.
+14. Do not use “__”.
+15. Use clear and explanatory variable names.
+
+## 3 -- Style Rules
+1. Do not indent pre-processor directives -- there should be no leading spaces before a directive.  (*)
+2. Never change the language syntax using **#deﬁne**.
+3. Do not use spaces between method names and their argument list e.g. `foo()` rather than
+`foo ()`.
+4. Do not use spaces in front of [], () and on either side of -> . For example, `vector[i]` instead of `vector [i]`
```

Comment:
Do we want to make an exception for the space immediately after builtin keywords like if/for/while? Either leave it optional or require the space? I personally usually add that space, although I've noticed others not adding them. I also noticed in the last attempted (but not approved yet) version of the automatic clang formatter those spaces are being added.

Reply:
We are not making a rule for if/for/while. The group made the decision early on that these rules would be brief and not cover every possible issue.
There will be a second phase where we look at implementing more rule checking with clang-tidy.

Reply:
What you are saying sounds good to me, but in that case I think the parentheses should just be eliminated from rule 4. You already covered the case of function calls in 3. If you like you could add to 3, "A similar rule applies to parentheses used in initialization." And if there is some other case you have in mind mention that. There are many other contexts where parentheses get used that you probably are not intending to include in 4, but taken literally it would apply to all of them.

Reply:
Sorry, I wasn't thinking through the implications. I dropped () from the list. I decided to add brace initialization with just a weak recommendation.